crates/storage/provider/src/traits/state.rs

```diff
@@ -12,68 +14,98 @@ pub type StateProviderBox = Box<dyn StateProvider>;
 
 /// An abstraction for a type that provides state data.
 #[auto_impl(&, Arc, Box)]
-pub trait StateProvider: BlockHashReader + AccountReader + StateRootProvider + Send + Sync {
+pub trait StateProvider:
+    BlockHashReader + AccountReader + BytecodeProvider + StateRootProvider + Send + Sync
+{
     /// Get storage of given account.
     fn storage(
         &self,
         account: Address,
         storage_key: StorageKey,
     ) -> ProviderResult<Option<StorageValue>>;
 
-    /// Get account code by its hash
-    fn bytecode_by_hash(&self, code_hash: B256) -> ProviderResult<Option<Bytecode>>;
+    /// Get account code by its address.
+    ///
+    /// Returns `None` if the account doesn't exist or account is not a contract
+    fn account_code(&self, address: Address) -> ProviderResult<Option<Bytecode>> {
+        // Get basic account information.
+        // Returns `None` if account doesn't exist.
+        let Some(account) = self.basic_account(address)? else { return Ok(None) };
+
+        if let Some(code_hash) = account.bytecode_hash.filter(|hash| hash != &KECCAK_EMPTY) {
+            self.bytecode_by_hash(code_hash)
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Get account balance by its address.
+    ///
+    /// Returns `None` if the account doesn't exist.
+    fn account_balance(&self, address: Address) -> ProviderResult<Option<U256>> {
+        Ok(self.basic_account(address)?.map(|a| a.balance))
+    }
+
+    /// Get account nonce by its address.
+    ///
+    /// Returns `None` if the account doesn't exist.
+    fn account_nonce(&self, address: Address) -> ProviderResult<Option<u64>> {
+        Ok(self.basic_account(address)?.map(|a| a.nonce))
+    }
 
     /// Get account and storage proofs.
     fn proof(&self, address: Address, keys: &[B256]) -> ProviderResult<AccountProof>;
+}
 
-    /// Get account code by its address.
+/// An abstraction for a type that provides state data for accounts by their hashed address.
+#[auto_impl(&, Arc, Box)]
```

Comment:
nice, didn't know of this crate. we have it as dep in a lot of crates, maybe we want to move it to the workspace manifest? out of scope nonetheless